### PR TITLE
Reduce code duplication in `units.format.Generic` subclasses

### PR DESCRIPTION
**Pull request by eerovaher**
### Description

The `Generic` subclasses contained identical implementations of `_to_decomposed_alternative()` and very similar implementations of `_validate_unit()`. Moving the implementations to the common parent class simplifies the code even if the parent class itself does not use these methods.

~Moving `_validate_unit()` required adding a new `class_name: str | None` argument to it. If the FITS formatter class would be named `FITS` instead of `Fits` then the argument would not be needed because `_validate_unit()` could then simply use `cls.__name__`, but changing the name of a class sounds like a big change and adding a new argument to a private method is a much less drastic solution.~ EDIT: the `Fits` class was renamed in a separate pull request.

The primary goal of this pull request is to simplify code by reducing duplication, but the new methods are also type annotated.

- [x] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Review by nstarman — COMMENTED** [review]
This looks good! I'm at a conference this week, so I won't have time for an in-depth review, so merge away if @mhvk approves.

File: astropy/units/format/generic.py
Comment (mhvk) [inline astropy/units/format/generic.py:605]: How about instead of adding a new parameter, we allow classes to define their own name? A bit like we do with `baseframe`. And then do something like `class_name = getattr(cls, "name", cls.__name__)`. Or even go the full hog and do:
```
def __init_subclass__(cls, **kwargs):
    if not hasattr(cls, "name"):
        cls.name = cls.__name__
```

File: astropy/units/format/generic.py
Comment (mhvk) [inline astropy/units/format/generic.py:639]: Hmm, looks like another followup!

File: astropy/units/format/vounit.py
Comment (mhvk) [inline astropy/units/format/vounit.py:147]: I think this part is unique to `VOUnit`, so should be kept.

Aside: very worried if there are no test failures with its removal!

**Review by mhvk — COMMENTED** [review]
Nice! One comment on how to deal with the name, and one worry about VOUnit.

File: astropy/units/format/generic.py
Comment (eerovaher) [reply astropy/units/format/generic.py:605]: The formatter classes already have a `name`, which is publicly used by the formatter registry: https://github.com/astropy/astropy/blob/3be7d5631a03c35811f0cfd16b3c41f670aa7724/astropy/units/format/base.py#L32-L38
It is the lowercase version of `cls.__name__`,  except for `LatexInline` which has an explicitly specified name for better readability: https://github.com/astropy/astropy/blob/3be7d5631a03c35811f0cfd16b3c41f670aa7724/astropy/units/format/latex.py#L83
I would find a second name attribute confusing.

File: astropy/units/format/vounit.py
Comment (eerovaher) [reply astropy/units/format/vounit.py:147]: This check has been moved to `VOUnit._validate_unit()`, you just have to keep in mind that in `_get_unit_name()` `unit` is a `UnitBase`, but in `_validate_unit()` it is a `str`, so we have check `_custom_units` keys instead of values.

The intended behavior is being tested: https://github.com/astropy/astropy/blob/3be7d5631a03c35811f0cfd16b3c41f670aa7724/astropy/units/tests/test_format.py#L756-L768

File: astropy/units/format/generic.py
Comment (eerovaher) [reply astropy/units/format/generic.py:639]: Removing the deprecated class looks like a good task for a first-time contributor.

File: astropy/units/format/generic.py
Comment (mhvk) [reply astropy/units/format/generic.py:605]: Ah, I was wondering about a name, but had forgotten we already had it! Maybe we should just use `cls.name` in the error messages? After all, that is how they are called by the user.

File: astropy/units/format/vounit.py
Comment (mhvk) [inline astropy/units/format/vounit.py:135]: Ah, now I see how this works. That also means the super with the class name is fairly natural here.

**Review by mhvk — APPROVED** [review]
So, with your comments it is clear my main worries are gone - maybe it is simplest in this refactoring PR to just do what you have here, which doesn't change anything. I'll approve now but wait with merging just in case you like my suggestion of using `cls.name` in the error messages better.

File: astropy/units/format/generic.py
Comment (eerovaher) [reply astropy/units/format/generic.py:605]: But the messages here refer to the OGIP, FITS or VOUnit standards, not to the `"ogip"`, `"fits"` or `"vounit"` formats. I would be more comfortable with renaming the `Fits` class to `FITS` so we could just use `cls.__name__` here.

Comment (mhvk) [reply astropy/units/format/generic.py:605]: Yes, on further thought, those classes are not used at all outside of the format piece; I think it is fine to rename, especially as the all-capital version really is more logical - we do the same for OGIP.

Comment (eerovaher) [reply astropy/units/format/generic.py:605]: It would be best to open a separate pull request for renaming the `Fits` class to `FITS` because it is an API change and shouldn't be done in a refactoring pull request.

I'll leave it to you to decide if we should put this pull request on hold until `Fits` has been renamed. The alternative would be to drop the second commit here, merge the first commit regarding `_to_decomposed_alternative()` as it is and then to open a third pull request for `_validate_unit()` after the `Fits` renaming pull request has been merged. It's getting late in my timezone, so if you think merging the first commit right away is better then you can just go ahead and do it.

**Review by mhvk — APPROVED** [review]
With the new FITS class, this is now all great!

File: astropy/units/format/generic.py
Comment (mhvk) [inline astropy/units/format/generic.py:603]: I've been looking at the format classes for the speed-up stuff, and one disadvantage of moving this method and the other one up here, is that they actually *fail* on `Generic` itself (since it has no `_units` attribute). That seems rather bad form. We may need an intermediate class, or think a bit more generally on how to refactor (perhaps not worrying about breaking backward compatibility for private methods).

Comment (eerovaher) [reply astropy/units/format/generic.py:603]: We have, over the last few months, found quite a lot of duplicated, dead or buggy code in the unit formatters. To me it seems that making small incremental changes to improve the situation has been the correct strategy even if a few individual updates have turned out to be suboptimal in hindsight. Regarding this change in particular, I will point out that at the time the `_units`, `_deprecated_units` and `_functions` class properties in `Generic` all depended on `_all_units`, which in turn depended on `_generate_unit_names()`, which was not implemented in `Generic`, not even as an abstract method. This means that even before this pull request there was already code in `Generic` that only its subclasses could use, so adding a couple more such methods was not a large qualitative change.

Given that the `astropy` units formatters are implemented as classes, it does make sense to use inheritance to reduce code duplication. But it is less clear if the base class for the parsing formatters should be a proper parsing formatter itself, not an abstract base class. Furthermore, the base class we have (i.e. `Generic`) has the only parser that is not limited to parsing ASCII, so the other formatters inherit quite a lot of Unicode-handling machinery that they don't use. It appears to me that the code would be simplified if the base class for parsing formatters would be less like `Generic` and more like `FITS`. That would be an API change, but major releases are the least bad time for such changes, so we might want to think about it.

Comment (mhvk) [reply astropy/units/format/generic.py:603]: Fair point that the methods didn't even work before! I very much agree with your choice of going for small, incremental changes to clean things up. But perhaps we are indeed reaching the limit of that - it seems also a bit odd that classes overwrite the parser, which presumably is the main thing that defines them.

Anyway, I don't really have an immediate solution, mostly just pointed this out since I think we are at a stage where we may need to define what we work towards. A possible solution also might be to actually define `_units` on `Generic` - after all, it is basically the registry.

Comment (eerovaher) [reply astropy/units/format/generic.py:603]: > I very much agree with your choice of going for small, incremental changes to clean things up. But perhaps we are indeed reaching the limit of that...

The idea that making small, incremental changes to the code makes the code quality asymptotically approach a local maximum sounds reasonable, as does the idea that this local maximum can differ from the global maximum. But the unit formatters are still quite far away from even the local maximum. I still see more opportunities for cleanup, but I'm avoiding opening multiple pull requests at a time because I want to avoid creating merge conflicts for myself.

Comment (mhvk) [reply astropy/units/format/generic.py:603]: Indeed! And happy to continue reviewing them!

**Pull request merged**
